test/harness/e2e/harness.go

```
@@ -75,60 +75,105 @@ func NewTestHarness() *Harness {
 	ctx, cancel := context.WithCancel(context.Background())
 
 	return &Harness{
-		VM:        testVM,
+		VMs:       []vm.TestVMInterface{testVM},
 		Client:    c,
 		Context:   ctx,
 		ctxCancel: cancel,
 		startTime: startTime,
 	}
 }
 
+func (h *Harness) DefaultVM() vm.TestVMInterface {
+    if len(h.VMs) == 0 {
+        Fail("No VMs available in the harness")
+    }
+    return h.VMs[0]
+}
+
+func (h *Harness) VM() vm.TestVMInterface {
+    return h.DefaultVM()
+}
+
+func (h *Harness) AddVM(vmParams vm.TestVM) vm.TestVMInterface {
+    testVM, err := vm.NewVM(vmParams)
+    Expect(err).ToNot(HaveOccurred())
+
+    h.VMs = append(h.VMs, testVM)
+    return testVM
+}
+
+func (h *Harness) AddMultipleVMs(vmParamsList []vm.TestVM) []vm.TestVMInterface {
+    var createdVMs []vm.TestVMInterface
+    for _, params := range vmParamsList {
+        createdVMs = append(createdVMs, h.AddVM(params))
+    }
+    return createdVMs
+}
+
+func (h *Harness) StopAllVMs() {
+    for _, vm := range h.VMs {
+        err := vm.Shutdown()
+        Expect(err).ToNot(HaveOccurred())
```

Same here, I would return the error, and validate that it doesn't happen in the test